packages/jupyter-ai/jupyter_ai/chat_handlers/tools.py

```diff
@@ -0,0 +1,260 @@
+        self.tools_file_path = None
+
+    # https://python.langchain.com/v0.2/docs/integrations/platforms/
+    def create_llm_chain(
+        self, provider: Type[BaseProvider], provider_params: Dict[str, str]
+    ):
+        unified_parameters = {
+            **provider_params,
+            **(self.get_model_parameters(provider, provider_params)),
+        }
+        llm = provider(**unified_parameters)
+        self.llm = llm
+        # self.chat_provider = self.setChatProvider(provider)
+        memory = ConversationBufferWindowMemory(
+            memory_key="chat_history", return_messages=True, k=2
+        )
+        self.llm_chain = LLMChain(
+            llm=self.llm, prompt=CONDENSE_PROMPT, memory=memory, verbose=False
+        )
+
+    # #### TOOLS FOR USE WITH LANGGRAPH #####
+    """
+    Bind tools to LLM and provide chat functionality.
+    Call:
+        /tools <query>
+    """
+
+    def conditional_continue(self, state: MessagesState) -> Literal["tools", "__end__"]:
+        messages = state["messages"]
+        last_message = messages[-1]
+        if last_message.tool_calls:
+            return "tools"
+        return "__end__"
+
+    # Get required tool files from ``.jupyter/jupyter-ai/tools/``
+    def getToolFiles(self, fpath):
+        if os.path.isfile(fpath):
+            file_paths = [fpath]
+        elif os.path.isdir(fpath):
+            file_paths = []
+            for filename in os.listdir(fpath):
+                file_paths.append(os.path.join(fpath, filename))
+        else:
+            self.reply("No tools found.")
+            return
+        return file_paths
+
+    def getToolNames(self, tools_file_path):
+        """
+        Read a file and extract the function names following the @tool decorator.
+        Args:
+            file_path (str): The path to the file.
+        Returns:
+            list: A list of function names.
+        """
+        try:
+            with open(tools_file_path) as file:
+                content = file.read()
+                tree = ast.parse(content)
+                tools = []
+                for node in ast.walk(tree):
+                    if isinstance(node, ast.FunctionDef):
+                        for decorator in node.decorator_list:
+                            if (
+                                isinstance(decorator, ast.Name)
+                                and decorator.id == "tool"
+                            ):
+                                tools.append(node.name)
+            return tools  # this is a list
+        except FileNotFoundError as e:  # to do
+            self.reply(f"Tools file not found at {tools_file_path}.")
+
+    def toolChat(self, query):
+        print("TOOL CHAT", query)
+        for chunk in self.app.stream(
+            {"messages": [("human", query)]}, stream_mode="values"
+        ):
+            response = chunk["messages"][-1].pretty_print()
+        return response
+
```

Comment:
This looks unused and not particularly helpful; can this be deleted?

Reply:
It was there in case we want to add a feature to stream the entire behind-the-scenes exchange between Human and AI but removing it as we can add it back later as an option.